### PR TITLE
Keyshortcuts

### PR DESCRIPTION
**Pull request by javerbukh**
Still in debug mode so there are many print statements. Currently working on finding where widgets and widget hierarchy are stored.

### CODE REVIEW

**Comment by astrofrog** [comment]
@javerbukh - it looks like you accidentally included the ``glue_core.egg-info`` directory - can you add a new commit that removes that directory? Thanks!

**Comment by javerbukh** [comment]
Hi Tom, how do I do that?

**Comment by astrofrog** [comment]
@javerbukh - you should be able to do:

```
git rm -r glue_core.egg-info
git commit -m "Removed glue_core.egg-info directory"
```

**Comment by javerbukh** [comment]
That gives me the error:

fatal: pathspec 'glue_core.egg_info' did not match any files

**Comment by astrofrog** [comment]
@javerbukh - could you maybe see if someone else locally could give you a hand with this? It would be good to also rebase to get rid of the merge commit. What is the current status of this at the moment - are you still having issues getting it to work in certain circumstances?

**Comment by javerbukh** [comment]
I have removed the debug statements and streamlined the code by a lot. Please let me know if I need to make any further changes and thank you for all of your help!

File: glue/app/qt/application.py [hunk 1]
```diff
@@ -36,7 +36,10 @@
 from glue.app.qt.plugin_manager import QtPluginManager
 from glue.app.qt.versions import QVersionsDialog
 from glue.app.qt.terminal import glue_terminal, IPythonTerminalError
-from glue.config import qt_fixed_layout_tab, qt_client, startup_action
+
+from datetime import datetime
```
Comment (astrofrog) [inline glue/app/qt/application.py]: Can you put this import nearer the top of the file? (normally standard library imports come first)

File: glue/app/qt/application.py [hunk 2]
```diff
@@ -205,6 +208,10 @@ def __init__(self, data_collection=None, session=None):
         Application.__init__(self, data_collection=data_collection,
                              session=session)
 
+        # Create eventFilter and pull in any keybindings from keyboard_shortcuts.py
```
Comment (astrofrog) [inline glue/app/qt/application.py]: I would remove the part about keyboard_shortcuts as there may be other shortcuts registered with ``keybindings``.

File: glue/app/qt/application.py [hunk 3]
```diff
@@ -362,6 +369,38 @@ def _setup_ui(self):
         self._log.resize(550, 550)
         self._log.hide()
 
+    def eventFilter(self, obj, event):
+        """
+        Mainly for handing Keyboard Shortcuts that are set
+        in keyboard_shortcuts.py
```
Comment (astrofrog) [inline glue/app/qt/application.py]: I would reword this - maybe just say ``Mainly for handling Keyboard Shortcuts`` and leave the rest out, since user-defined shortcuts will get picked up too.

File: glue/app/qt/application.py [hunk 4]
```diff
@@ -362,6 +369,38 @@ def _setup_ui(self):
...
+            if self.current_tab.activeSubWindow() and self.current_tab.activeSubWindow().widget():
+                active_window = self.current_tab.activeSubWindow().widget()
+            else:
+                active_window = None
+
+            # keybindings is a data structure in the form of dict(dict)
+            # which uses the DataViewer as the first key, the key pressed
+            # as the second key, and the function associated with those two
+            # as the value.
+            if self.keybindings.__contains__(str(active_window)):
```
Comment (astrofrog) [inline glue/app/qt/application.py]: This can be more nicely written as ``str(active_window) in self.keybindings``

File: glue/app/qt/application.py [hunk 5]
```diff
@@ -362,6 +369,38 @@ def _setup_ui(self):
...
+                for k, func in self.keybindings.members[str(active_window)].items():
+                    if event.key() == k:
```
Comment (astrofrog) [inline glue/app/qt/application.py]: Maybe use ``key`` instead of ``k`` for readability

File: glue/app/qt/application.py [hunk 6]
```diff
@@ -362,6 +369,38 @@ def _setup_ui(self):
...
+                        func(self.session)
+                        return True
+
+            # If key does not correspond with viewers, it might correspond
+            # with the global application, thus, None
+            if self.keybindings.__contains__("None"):
```
Comment (astrofrog) [inline glue/app/qt/application.py]: ``"None" in self.keybindings``

File: glue/config.py [hunk 7]
```diff
@@ -655,6 +654,40 @@ def default_members(self):
...
+
+    def add(self, valid_viewers, keybind, function):
+        """
+        Add a new keyboard shortcut
+        :param valid_viewers: list of viewers where event can be fired
+        :type label: list
+        :param keybind: type of key event
+        :type Qt.Key
+        :param function: function
+        :type function: function()
```
Comment (astrofrog) [inline glue/config.py]: Can you use the Numpydoc format for the docstring? (I'm trying to transition away from the sphinx format)

File: glue/config.py [hunk 8]
```diff
@@ -684,6 +717,7 @@ def __call__(self, state=None):
 menubar_plugin = MenubarPluginRegistry()
 preference_panes = PreferencePanesRegistry()
 qglue_parser = QGlueParserRegistry()
+keybindings = KeyboardShortcut()
```
Comment (astrofrog) [inline glue/config.py]: Maybe use ``keyboard_shortcut`` instead of ``keybindings``

**Comment by javerbukh** [comment]
Tom, to use the viewer classes themselves as keys rather than strings, I would have to change the decorators in keyboard_shortcuts.py. The decorators use a list of valid viewers (as strings) to determine when a shortcut can be used. Should I change this to a list of class instances? Can you give me an example of how that would look? (i.e. instanceof(2D image))

File: glue/keyboard_shortcuts.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    list_of_shortcuts = []
+    for k in viewer_tool.__iter__():
+        list_of_shortcuts.append(viewer_tool.members[k].shortcut)
+
+    if key_shortcut in list_of_shortcuts:
+        return True
+    return False
+
+
+@keybindings(QtCore.Qt.Key_Tab, ["2D Image", "1D Histogram", "2D Scatter", "Table Viewer"])
```
Comment (astrofrog) [inline glue/keyboard_shortcuts.py]: Here you could list the classes directly for the viewers (you'll need to import them into this file)

File: glue/app/qt/application.py [hunk 10]
```diff
@@ -362,6 +368,37 @@ def _setup_ui(self):
...
+                active_window = self.current_tab.activeSubWindow().widget()
+            else:
+                active_window = None
+
+            # keybindings is a data structure in the form of dict(dict)
+            # which uses the DataViewer as the first key, the key pressed
+            # as the second key, and the function associated with those two
+            # as the value.
+            if str(active_window) in self.keybindings:
+                for k, func in self.keybindings.members[str(active_window)].items():
```
Comment (astrofrog) [inline glue/app/qt/application.py]: In the two lines above you could change ``str`` to ``type``

File: glue/app/qt/application.py [hunk 11]
```diff
@@ -362,6 +368,37 @@ def _setup_ui(self):
...
+                    if event.key() == k:
+                        func(self.session)
+                        return True
+
+            # If key does not correspond with viewers, it might correspond
+            # with the global application, thus, None
+            if "None" in self.keybindings:
+                for k, func in self.keybindings.members["None"].items():
```
Comment (astrofrog) [inline glue/app/qt/application.py]: Here you could change ``"None"`` to ``None``

**Comment by astrofrog** [comment]
@javerbukh - before we can merge this, could you do a rebase (no need to squash, just plain rebase) to get rid of the merge commits?

I'm a little worried that the CI builds are failing due to timeout - this suggests that the installed event filter is slowing things down significantly. I need to investigate this before pressing the merge button

**Comment by javerbukh** [comment]
I finished the rebase. I'm not sure if I need to do anything after it's completed

**Comment by astrofrog** [comment]
@javerbukh - you need to force push your changes to this branch

**Comment by javerbukh** [comment]
Done

**Comment by drdavella** [comment]
@astrofrog, @javerbukh: we just tested this branch locally and noticed no difference in test timing compared to `master`. Did you observe meaningful timing differences somewhere else? Is it possible that the test timeouts are unrelated?

In any case, I'm not necessarily opposed to having to explicitly enable event handling for keyboard shortcuts since we can easily handle this on the application side of things.

</unsolicited opinion>
cc @brechmos-stsci

**Comment by astrofrog** [comment]
@javerbukh - I pushed a small change to move ``keyboard_shortcuts.py`` to ``glue/app/qt`` since it is currently Qt-specific. I will go ahead and merge this, and will open an issue with some thoughts on further work that could be done in terms of keyboard shortcuts.

**Comment by astrofrog** [comment]
@javerbukh - thanks for working on this!
